File: frontend/src/lib/ky-tieu-chi.ts
/**
 * src/lib/ky-tieu-chi.ts
 * ======================
 * Kỳ chấm TIÊU CHÍ CHUNG (30 điểm): theo THÁNG (cũ) hay theo QUÝ (mới).
 *
 * Căn cứ: Công văn 21169/CHQ-TCCB ngày 28/8/2026 — "Từ quý III/2026, kỳ đánh giá,
 * xếp loại trên Phần mềm được thiết lập theo quý".
 *
 * PHẢI giữ đồng bộ với backend `app/core/ky_tieu_chi.py` (cùng mốc, cùng quy tắc
 * tháng neo). Backend luôn tự quy tháng gửi lên về tháng neo, nên phía FE chỉ cần
 * đúng để HIỂN THỊ và dựng bộ chọn kỳ.
 */

/** (năm, quý) — mọi kỳ TỪ mốc này trở đi chấm tiêu chí chung theo QUÝ. */
export const TC_THEO_QUY_TU = { nam: 2026, quy: 3 };

export function quyCuaThang(thang: number): number {
  return Math.floor((thang - 1) / 3) + 1;
}

export function cacThangTrongQuy(quy: number): number[] {
  const dau = (quy - 1) * 3 + 1;
  return [dau, dau + 1, dau + 2];
}

export function thangCuoiQuy(quy: number): number {
  return quy * 3;
}

/** Kỳ này chấm tiêu chí chung theo quý chưa? */
export function tcTheoQuy(thang: number, nam: number): boolean {
  const quy = quyCuaThang(thang);
  return nam > TC_THEO_QUY_TU.nam || (nam === TC_THEO_QUY_TU.nam && quy >= TC_THEO_QUY_TU.quy);
}

/** Tháng chứa phiếu tiêu chí của kỳ (kỳ quý → tháng cuối quý). */
export function thangNeo(thang: number, nam: number): number {
  return tcTheoQuy(thang, nam) ? thangCuoiQuy(quyCuaThang(thang)) : thang;
}

/** Các tháng dùng chung điểm tiêu chí của kỳ. */
export function cacThangApDung(thang: number, nam: number): number[] {
  return tcTheoQuy(thang, nam) ? cacThangTrongQuy(quyCuaThang(thang)) : [thang];
}

/** Nhãn hiển thị: 'Quý 3/2026' hoặc 'Tháng 5/2026'. */
export function nhanKy(thang: number, nam: number): string {
  return tcTheoQuy(thang, nam) ? `Quý ${quyCuaThang(thang)}/${nam}` : `Tháng ${thang}/${nam}`;
}

/** Nhãn ngắn cho bộ chọn: 'Quý 3' hoặc 'Tháng 5'. */
export function nhanKyNgan(thang: number, nam: number): string {
  return tcTheoQuy(thang, nam) ? `Quý ${quyCuaThang(thang)}` : `Tháng ${thang}`;
}

/**
 * Kỳ đã bắt đầu chưa. Kỳ quý neo ở tháng cuối quý nên tháng neo có thể ở tương
 * lai trong khi quý đã bắt đầu (chấm Quý IV từ tháng 10, phiếu neo ở T12).
 */
export function kyDaBatDau(thang: number, nam: number, homNay: Date = new Date()): boolean {
  const namNay = homNay.getFullYear();
  const thangNay = homNay.getMonth() + 1;
  if (tcTheoQuy(thang, nam)) {
    const quy = quyCuaThang(thang);
    return nam < namNay || (nam === namNay && quy <= quyCuaThang(thangNay));
  }
  return nam < namNay || (nam === namNay && thang <= thangNay);
}

/** Kỳ đánh giá, xếp loại THÁNG còn hiệu lực với tháng/năm này không. */
export function kyThangConHieuLuc(thang: number, nam: number): boolean {
  return !tcTheoQuy(thang, nam);
}

export interface TuyChonKy {
  /** Giá trị gửi lên backend — luôn là một số tháng (kỳ quý → tháng neo). */
  thang: number;
  nhan: string;
  laQuy: boolean;
}

/**
 * Danh sách kỳ chọn được trong một năm: tháng cho kỳ cũ, quý cho kỳ mới.
 * Ví dụ 2026 → Tháng 1..6 + Quý 3 + Quý 4; 2027 → Quý 1..4.
 */
export function danhSachKyTrongNam(nam: number): TuyChonKy[] {
  const ds: TuyChonKy[] = [];
  const quyDaThem = new Set<number>();
  for (let thang = 1; thang <= 12; thang++) {
    if (!tcTheoQuy(thang, nam)) {
      ds.push({ thang, nhan: `Tháng ${thang}`, laQuy: false });
      continue;
    }
    const quy = quyCuaThang(thang);
    if (quyDaThem.has(quy)) continue;
    quyDaThem.add(quy);
    ds.push({ thang: thangCuoiQuy(quy), nhan: `Quý ${quy}`, laQuy: true });
  }
  return ds;
}

// =============================================================================
// XEM LẠI KỲ THÁNG (22/09/2026)
// -----------------------------------------------------------------------------
// Trang Đánh giá và trang Tiêu chí chung vẫn cần chọn được các THÁNG đã đánh giá
// xong theo kỳ tháng, kể cả tháng đã nằm trong một quý. Mốc dưới là tháng cuối
// cùng thực sự chấm xong theo tháng: tháng 7/2026 có 517 đơn tiêu chí đã duyệt,
// tháng 8 chỉ chấm dở (322/458) rồi dừng khi chuyển sang kỳ quý.
// =============================================================================

/** Tháng cuối cùng còn hiện thành một mục "Tháng" trong bộ chọn kỳ. */
export const THANG_XEM_LAI_DEN = { nam: 2026, thang: 7 };

/** Tháng này còn được chọn riêng để XEM LẠI số liệu đã chấm theo tháng không? */
export function conXemLaiTheoThang(thang: number, nam: number): boolean {
  return (
    nam < THANG_XEM_LAI_DEN.nam ||
    (nam === THANG_XEM_LAI_DEN.nam && thang <= THANG_XEM_LAI_DEN.thang)
  );
}

/**
 * Tháng đang chọn là SỐ LIỆU LỊCH SỬ: đã thuộc một kỳ quý nhưng vẫn cho xem lại
 * theo tháng. Màn hình phải để chế độ chỉ đọc — mọi thao tác GHI đều bị backend
 * quy về phiếu quý, nếu để form mở thì người dùng tưởng sửa tháng 7 mà thực ra
 * đang ghi đè phiếu Quý III.
 */
export function laThangLichSu(thang: number, nam: number): boolean {
  return tcTheoQuy(thang, nam) && conXemLaiTheoThang(thang, nam);
}

/**
 * Danh sách kỳ cho trang ĐÁNH GIÁ và trang TIÊU CHÍ CHUNG: các tháng còn xem lại
 * được, rồi tới các quý. Năm 2026 ra: Tháng 1…7 · Quý III · Quý IV.
 */
export function danhSachKyXemLai(nam: number): TuyChonKy[] {
  const ds: TuyChonKy[] = [];
  const quyDaThem = new Set<number>();
  for (let thang = 1; thang <= 12; thang++) {
    if (conXemLaiTheoThang(thang, nam)) {
      ds.push({ thang, nhan: `Tháng ${thang}`, laQuy: false });
    }
    if (!tcTheoQuy(thang, nam)) continue;
    const quy = quyCuaThang(thang);
    if (quyDaThem.has(quy)) continue;
    quyDaThem.add(quy);
    ds.push({ thang: thangCuoiQuy(quy), nhan: `Quý ${quy}`, laQuy: true });
  }
  return ds;
}

/**
 * Kỳ mặc định khi mở trang: quý hiện hành nếu năm đó đã chuyển sang kỳ quý,
 * ngược lại là tháng hiện tại (giữ hành vi cũ cho các năm trước).
 */
export function kyMacDinh(nam: number, homNay: Date = new Date()): TuyChonKy {
  const ds = danhSachKyXemLai(nam);
  const thangNay = homNay.getMonth() + 1;
  const thangMoc = nam === homNay.getFullYear() ? thangNay : 12;

  if (tcTheoQuy(thangMoc, nam)) {
    const quy = quyCuaThang(thangMoc);
    const muc = ds.find((k) => k.laQuy && k.thang === thangCuoiQuy(quy));
    if (muc) return muc;
  }
  const mucThang = ds.find((k) => !k.laQuy && k.thang === thangMoc);
  return mucThang ?? ds[ds.length - 1];
}

File: backend/app/core/ky_tieu_chi.py
"""
app/core/ky_tieu_chi.py
=======================
Kỳ chấm TIÊU CHÍ CHUNG (30 điểm): theo THÁNG (cũ) hay theo QUÝ (mới).

Căn cứ: Công văn 21169/CHQ-TCCB ngày 28/8/2026 của Cục Hải quan và Công văn
triển khai của Chi cục Hải quan khu vực VIII — "Từ quý III/2026, kỳ đánh giá,
xếp loại trên Phần mềm được thiết lập theo quý".

CÁCH LÀM — "phiếu tiêu chí của quý neo ở THÁNG CUỐI QUÝ":
    Từ kỳ >= TC_THEO_QUY_TU, việc chấm và duyệt tiêu chí chung chỉ diễn ra trên
    MỘT bản ghi `danh_gia_thang` — bản ghi của tháng cuối quý (T3, T6, T9, T12),
    gọi là THÁNG NEO, đánh dấu bằng cờ `danh_gia_thang.la_phieu_tc_quy`.
    Hai tháng còn lại của quý không chấm nữa; mọi nơi cần điểm/chi tiết tiêu chí
    của chúng đọc xuyên sang bản ghi neo qua các hàm trong module này.

    Ví dụ Quý III/2026: T7 và T8 đọc sang bản ghi T9.

NGUYÊN TẮC AN TOÀN:
    Chỉ đổi nơi ĐỌC. KHÔNG ghi đè điểm/trạng thái tiêu chí của các tháng cũ —
    dữ liệu đã chấm theo tháng (kể cả T7, T8/2026) giữ nguyên để tra cứu.

QUAY LUI:
    Đặt TC_THEO_QUY_TU = (9999, 1) rồi phát hành lại → toàn hệ thống trở về chấm
    theo tháng ngay lập tức, dữ liệu phiếu quý nằm im, không cần migration ngược.

Phiên bản: 1.0.0 (18/09/2026)
"""

from datetime import date, timedelta
from typing import Dict, List, Optional, Tuple

# (năm, quý) — mọi kỳ TỪ mốc này trở đi chấm tiêu chí chung theo QUÝ.
# Quý III/2026 theo Công văn 21169.
TC_THEO_QUY_TU: Tuple[int, int] = (2026, 3)

QUY_TO_THANG = {1: [1, 2, 3], 2: [4, 5, 6], 3: [7, 8, 9], 4: [10, 11, 12]}


def quy_cua_thang(thang: int) -> int:
    """Quý chứa tháng: 7 → 3."""
    return (int(thang) - 1) // 3 + 1


def cac_thang_trong_quy(quy: int) -> List[int]:
    """Ba tháng của quý: 3 → [7, 8, 9]."""
    return list(QUY_TO_THANG.get(int(quy), []))


def thang_cuoi_quy(quy: int) -> int:
    """Tháng cuối quý: 3 → 9."""
    return int(quy) * 3


def tc_theo_quy(thang: int, nam: int) -> bool:
    """True nếu kỳ (tháng/năm) này chấm tiêu chí chung THEO QUÝ."""
    return (int(nam), quy_cua_thang(thang)) >= TC_THEO_QUY_TU


def tc_theo_quy_cua_quy(quy: int, nam: int) -> bool:
    """Bản dùng trực tiếp với (quý, năm) — cho các luồng cấp quý."""
    return (int(nam), int(quy)) >= TC_THEO_QUY_TU


def thang_neo(thang: int, nam: int) -> int:
    """
    Tháng chứa phiếu tiêu chí của kỳ.

    - Kỳ theo quý  → tháng cuối quý (7, 8, 9 → 9).
    - Kỳ theo tháng → chính tháng đó (giữ nguyên hành vi cũ).
    """
    if tc_theo_quy(thang, nam):
        return thang_cuoi_quy(quy_cua_thang(thang))
    return int(thang)


def la_thang_neo(thang: int, nam: int) -> bool:
    """True nếu bản ghi tháng này là nơi chấm tiêu chí của kỳ."""
    return int(thang) == thang_neo(thang, nam)


def cac_thang_ap_dung(thang: int, nam: int) -> List[int]:
    """Các tháng dùng chung điểm tiêu chí của kỳ (kỳ tháng → chỉ chính nó)."""
    if tc_theo_quy(thang, nam):
        return cac_thang_trong_quy(quy_cua_thang(thang))
    return [int(thang)]


def nhan_ky(thang: int, nam: int) -> str:
    """Nhãn hiển thị: 'Quý 3/2026' hoặc 'Tháng 5/2026'."""
    if tc_theo_quy(thang, nam):
        return f"Quý {quy_cua_thang(thang)}/{nam}"
    return f"Tháng {thang}/{nam}"


def loai_ky(thang: int, nam: int) -> str:
    """'QUY' hoặc 'THANG' — FE dùng để đổi bộ chọn kỳ."""
    return "QUY" if tc_theo_quy(thang, nam) else "THANG"


def thong_tin_ky(thang: int, nam: int, theo_dung_thang: bool = False) -> dict:
    """
    Gói thông tin kỳ trả về cho FE (đính kèm mọi response tiêu chí chung).

    `theo_dung_thang=True` — chỉ dùng cho đường ĐỌC LỊCH SỬ: người dùng xem lại
    điểm tiêu chí đã chấm theo tháng trước khi công văn 21169 có hiệu lực. Khi đó
    loại kỳ là THANG_LICH_SU để giao diện dán nhãn "số liệu lịch sử", tránh hiểu
    nhầm đây là điểm đang có hiệu lực.
    """
    if theo_dung_thang and tc_theo_quy(thang, nam):
        return {
            "ky": "THANG_LICH_SU",
            "quy": quy_cua_thang(thang),
            "thang_neo": int(thang),
            "cac_thang_ap_dung": [int(thang)],
            "nhan_ky": f"Tháng {thang}/{nam} (số liệu lịch sử)",
        }
    return {
        "ky": loai_ky(thang, nam),
        "quy": quy_cua_thang(thang) if tc_theo_quy(thang, nam) else None,
        "thang_neo": thang_neo(thang, nam),
        "cac_thang_ap_dung": cac_thang_ap_dung(thang, nam),
        "nhan_ky": nhan_ky(thang, nam),
    }


def ky_thang_con_hieu_luc(thang: int, nam: int) -> bool:
    """
    Kỳ đánh giá, xếp loại THÁNG còn hiệu lực với tháng/năm này không.

    CV 21169 thay thế các văn bản hướng dẫn đánh giá hằng tháng của Chi cục, và
    "từ quý III/2026 kỳ đánh giá, xếp loại trên Phần mềm được thiết lập theo quý"
    → từ mốc TC_THEO_QUY_TU, KHÔNG lập mới báo cáo/phiếu đánh giá theo tháng nữa.
    Dữ liệu tháng đã có vẫn đọc và in được bình thường.
    """
    return not tc_theo_quy(thang, nam)


def ky_da_bat_dau(thang: int, nam: int, hom_nay: Optional[date] = None) -> bool:
    """
    Kỳ đã bắt đầu hay chưa (dùng thay cho phép so sánh tháng <= hôm nay).

    Với kỳ theo quý, phiếu neo ở THÁNG CUỐI quý nên tháng neo có thể nằm ở
    tương lai trong khi quý đã bắt đầu (ví dụ chấm Quý IV vào tháng 10, phiếu
    neo ở T12). Khi đó vẫn phải cho chấm.
    """
    hom_nay = hom_nay or date.today()
    if tc_theo_quy(thang, nam):
        return (int(nam), quy_cua_thang(thang)) <= (
            hom_nay.year,
            quy_cua_thang(hom_nay.month),
        )
    return (int(nam), int(thang)) <= (hom_nay.year, hom_nay.month)


# =============================================================================
# MỐC CHUYỂN KỲ CỦA KÊ KHAI CÔNG VIỆC (22/09/2026)
# -----------------------------------------------------------------------------
# Hồ sơ đánh giá quý phải nộp ngày 23 của tháng cuối quý (CV 21169), nên công
# việc làm sau mốc chốt số liệu không kịp vào hồ sơ quý đó. Quyết định của Chi
# cục: kê khai có NGÀY THỰC HIỆN từ 16/9/2026 trở đi tính vào quý IV.
#
# Phạm vi hẹp, cố ý:
#   • CHỈ áp cho phần điểm tính từ KÊ KHAI CÔNG VIỆC (a/b/c) — cũng là phần
#     điểm lãnh đạo cộng SP cấp dưới.
#   • KHÔNG đụng điểm THÁNG: tháng 9 vẫn tính trọn 1–30/9 để tra cứu.
#   • KHÔNG đụng d/đ/e, tiêu chí chung, hay HĐLĐ 111 (điểm từ VB714 theo tháng).
#   • Chỉ một lần cho Q3→Q4/2026; các quý sau vẫn chia theo tháng như cũ.
#
# Bản ghi THIẾU ngày thực hiện (16 bản của T9/2026) được giữ ở quý GỐC — nếu lọc
# cứng theo ngày thì chúng rơi khỏi cả hai quý và biến mất khỏi mọi bảng điểm.
# =============================================================================

# (năm, quý) → ngày đầu tiên thuộc quý KẾ TIẾP
MOC_CHUYEN_KY_KE_KHAI: dict = {
    (2026, 3): date(2026, 9, 16),
}


def _moc_cua_quy(quy: int, nam: int) -> Optional[date]:
    return MOC_CHUYEN_KY_KE_KHAI.get((int(nam), int(quy)))


def cac_thang_ke_khai_cua_quy(quy: int, nam: int) -> List[Tuple[int, Optional[date], Optional[date]]]:
    """
    Các tháng đóng góp KÊ KHAI cho quý, kèm cửa sổ ngày (tu_ngay, den_ngay).

    None = không giới hạn. Ví dụ năm 2026:
        quý 3 → [(7, None, None), (8, None, None), (9, None, 15/9)]
        quý 4 → [(9, 16/9, None), (10, None, None), (11, None, None), (12, None, None)]
    """
    ds: List[Tuple[int, Optional[date], Optional[date]]] = []

    # Phần đuôi của quý trước bị đẩy sang quý này
    moc_truoc = _moc_cua_quy(quy - 1, nam) if quy > 1 else None
    if moc_truoc is not None:
        ds.append((thang_cuoi_quy(quy - 1), moc_truoc, None))

    moc = _moc_cua_quy(quy, nam)
    for thang in cac_thang_trong_quy(quy):
        if moc is not None and thang == thang_cuoi_quy(quy):
            # Tháng cuối quý chỉ tính đến hôm trước mốc
            ds.append((thang, None, moc - timedelta(days=1)))
        else:
            ds.append((thang, None, None))
    return ds


def co_moc_chuyen_ky(quy: int, nam: int) -> bool:
    """Quý này có dính mốc chuyển kỳ (của chính nó hoặc của quý trước) không."""
    return _moc_cua_quy(quy, nam) is not None or (
        quy > 1 and _moc_cua_quy(quy - 1, nam) is not None
    )
